src/routes/solid-start/advanced/middleware.mdx

```diff
@@ -2,28 +2,227 @@
+						"Access-Control-Allow-Methods": "POST, PUT, DELETE",
+						"Access-Control-Allow-Headers": "Content-Type, Authorization",
+					},
+				});
+			}
+
+			// Handle normal requests
+			event.response.headers.set("Access-Control-Allow-Origin", origin);
+		}
+	},
 });
 ```
 
-Inside the middleware file, you can export a `createMiddleware` function.
+## Using cookies
+
+Cookies are regular headers. On a request, they are stored in the `Cookie` header. On a response they are in the `Set-Cookie` header.
+
+Vinxi provides helpers to work with cookies more easily.
 
-```tsx
+```ts
 import { createMiddleware } from "@solidjs/start/middleware";
+import { getCookie, setCookie } from "vinxi/http";
 
 export default createMiddleware({
-  onRequest: [
-    event => {
-      console.log("GLOBAL", event.request.url);
-    }
-  ]
+	onRequest: (event) => {
+		// Get a cookie
+		const theme = getCookie(event.nativeEvent, "theme");
+
+		// Set a new cookie
+		setCookie("affiliate", "my-affiliate-id");
+	},
 });
 ```
 
-Middleware supports 2 lifecycles: `onRequest` and `onBeforeResponse`. If you return a value from middleware it will respond with that, otherwise it will run the next one in the chain.
\ No newline at end of file
+You can [read more about Vinxi helpers here](https://vinxi.vercel.app/api/server/cookies.html).
+
+## Chaining middlewares
+
+Both `onRequest` and `onBeforeResponse` event handlers can accept an array of middleware functions. In this case, all middleware will be executed in order.
+
+```ts
+import { createMiddleware } from "@solidjs/start/middleware";
+import { type FetchEvent } from "@solidjs/start/server";
+
+function middleware1(event: FetchEvent) {
+	event.request.headers.set("x-custom-header1", "hello-from-middleware1");
+}
+
+function middleware2(event: FetchEvent) {
+	event.request.headers.set("x-custom-header2", "hello-from-middleware2");
+}
+
+export default createMiddleware({
+	onRequest: [middleware1, middleware2],
+});
+```
+
+## Redirects
+
+Solid-Router provides a `redirect` helper that can be used to redirect the user to a different URL within a middleware.
+
+```ts
+import { createMiddleware } from "@solidjs/start/middleware";
+import { redirect } from "@solidjs/router";
+
+const REDIRECT_MAP: Record<string, string> = {
+	"/signup": "/auth/signup",
+	"/login": "/auth/login",
+};
+
+export default createMiddleware({
+	onRequest: (event) => {
+		const { pathname } = new URL(event.request.url);
+
+		if (pathname in REDIRECT_MAP) {
+			return redirect(REDIRECT_MAP[pathname], 301);
+		}
+	},
+});
+```
+
+## Common Pitfalls
+
+### Authentication and Authorization
+
+Although using middleware for authentication and authorization is a common practice in many web frameworks, it is not advisable to use SolidStart middleware for these purposes.
+
+This is because SolidStart middleware is not guaranteed to run on every single request.
+
+For optimal security, all checks should be performed as close to your data source as possible.
```

Comment:
You can actually add a section on Authentication that shows how to do token validation in a middleware, and reword this to something like this

> Using middleware for protecting routes is not recommended in SolidStart. Authorization should be handled as close as possible to the data source.

Technically authentication is completely fine (and recommended) in the middleware, authorization is the thing that should be close to the data and not in the middleware.